app/src/main/java/org/simple/clinic/ReplayUntilScreenIsDestroyed.kt

```diff
@@ -0,0 +1,45 @@
+package org.simple.clinic
+
+import io.reactivex.Observable
+import io.reactivex.ObservableSource
+import io.reactivex.ObservableTransformer
+import io.reactivex.disposables.CompositeDisposable
+import io.reactivex.rxkotlin.ofType
+import io.reactivex.rxkotlin.plusAssign
+import org.simple.clinic.widgets.ScreenDestroyed
+import org.simple.clinic.widgets.UiEvent
+
+class ReplayUntilScreenIsDestroyed(
+    private val transforms: List<ObservableTransformer<UiEvent, UiEvent>> = emptyList()
+) : ObservableTransformer<UiEvent, UiEvent> {
+
+  constructor(transform: ObservableTransformer<UiEvent, UiEvent>) : this(transforms = listOf(transform))
+
+  constructor(
+      transform1: ObservableTransformer<UiEvent, UiEvent>,
+      transform2: ObservableTransformer<UiEvent, UiEvent>
+  ) : this(transforms = listOf(transform1, transform2))
+
+  private val disposables = CompositeDisposable()
+
+  override fun apply(upstream: Observable<UiEvent>): ObservableSource<UiEvent> {
+    val replayedEvents = upstream
+        .replay()
+        .autoConnect(1) { disposables += it }
+
+    val transformedEvents = transforms
+        .fold(replayedEvents) { events, transform ->
+          events.compose(transform)
+              .replay()
+              .autoConnect(1) { disposables += it }
+        }
+
+    disposeWhenScreenIsDestroyed(transformedEvents)
+
+    return transformedEvents
+  }
+
+  private fun disposeWhenScreenIsDestroyed(events: Observable<UiEvent>) {
+    disposables.add(events.ofType<ScreenDestroyed>().subscribe { disposables.clear() })
```

Comment:
This is slightly compact because both `disposables.add()` and `disposables.clear()` are happening on the same line. Can we break this line?

```kotlin
disposables.add(
    events
        .ofType<ScreenDestroyed>()
        .subscribe { disposables.clear() })
```

Reply:
Yup, I'll do this.